helix-term/src/commands.rs

```diff
@@ -5100,3 +5103,93 @@ fn rename_symbol(cx: &mut Context) {
+    increment_number_impl(cx, cx.count() as i64);
+}
+
+/// If there is a number at the cursor, decrement it by the count.
+fn decrement_number(cx: &mut Context) {
+    increment_number_impl(cx, -(cx.count() as i64));
+}
+
+/// If there is a number at the cursor, increment it by `amount`.
+fn increment_number_impl(cx: &mut Context, amount: i64) {
+    let (view, doc) = current!(cx.editor);
+    let selection = doc.selection(view.id);
+    let text = doc.text();
+
+    let changes = selection.ranges().iter().filter_map(|range| {
+        if let Some(NumberInfo {
+            range,
+            value: old_value,
+            radix,
+        }) = number_at(text.slice(..), *range)
+        {
+            let new_value = old_value.wrapping_add(amount);
+            let old_text: Cow<str> = text.slice(range.from()..range.to()).into();
+            let prefix = if radix == 10 { "" } else { &old_text[..2] };
+
+            let mut new_text = match radix {
+                2 => format!("{:b}", new_value),
+                8 => format!("{:o}", new_value),
+                10 => format!("{}", new_value.abs()),
+                16 => {
+                    let lower_count = old_text
+                        .chars()
+                        .skip(2)
+                        .filter(char::is_ascii_lowercase)
+                        .count();
+                    let upper_count = old_text
+                        .chars()
+                        .skip(2)
+                        .filter(char::is_ascii_uppercase)
+                        .count();
+                    if upper_count > lower_count {
+                        format!("{:X}", new_value)
+                    } else {
+                        format!("{:x}", new_value)
+                    }
+                }
+                _ => unimplemented!("radix not supported: {}", radix),
+            };
+
+            // Pad with leading zeros if necessary.
+            // * For non-decimal numbers, we want to keep at least as many digits as before the change.
+            // * For decimal numbers we do this if there are leading zeros, like 000145, before the change.
+            let new_length_no_prefix = new_text.len();
+            let old_text_no_prefix = {
+                let mut stripped: &str = old_text[prefix.len()..].as_ref();
+                if stripped.starts_with('-') {
+                    stripped = &stripped[1..];
+                }
+                stripped
+            };
+            let old_length_no_prefix = old_text_no_prefix.len();
+            if new_length_no_prefix < old_length_no_prefix
+                && (radix != 10 || old_text_no_prefix.starts_with('0'))
+            {
+                new_text = "0".repeat(old_length_no_prefix - new_length_no_prefix) + &new_text;
+            }
+
+            // Add prefix or sign if needed
+            if radix == 10 && new_value.is_negative() {
+                new_text = format!("-{}", new_text);
```

Comment:
I wonder, couldn't you remove `.abs()` earlier in the `10 =>` branch and it'll add the sign for you already?

Reply:
The reason the `-` is added later is because of the possible need to pad with leading zeros so that `-00005` gets incremented to `-00004`.

Reply:
Ah I see. You can do both padding and add the sign in a single format:

```rust
format!("{:01$}",  -101isize, 5) // -101 padded to 5 characters
```

note that it includes the sign in the length, so you get `-0101`

```rust
            let mut new_text = match radix {
                2 => format!("{:01$b}", new_value, old_length),
                8 => format!("{:01$o}", new_value, old_length),
                10 => format!("{01$}", new_value, old_length),
```

Reply:
Good to know. That's much better!